src/MongoDB.EntityFrameworkCore/ChangeTracking/CollectionComparer.cs

```diff
@@ -0,0 +1,129 @@
+    {
+        if (ReferenceEquals(a, b)) return true;
+        if (a is null) return b is null;
+        if (b is null) return false;
+
+        if (a is IList<TElement> aList && b is IList<TElement> bList)
+        {
+            if (aList.Count != bList.Count) return false;
+
+            for (int i = 0; i < aList.Count; i++)
+            {
+                var (el1, el2) = (aList[i], bList[i]);
+                if (el1 is null)
+                {
+                    if (el2 is null) continue;
+                    return false;
+                }
+
+                if (el2 is null) return false;
+                if (!elementComparer.Equals(el1, el2)) return false;
+            }
+
+            return true;
+        }
+
+        throw new InvalidOperationException(
+            CoreStrings.BadListType(
+                (a is IList<TElement?> ? b : a).GetType().ShortDisplayName(),
+                typeof(IList<>).MakeGenericType(elementComparer.Type).ShortDisplayName()));
+    }
+
+    private static int GetHashCode(IEnumerable<TElement> source, ValueComparer<TElement> elementComparer)
+    {
+        var hash = new HashCode();
+        foreach (var element in source)
+        {
+            hash.Add(element, elementComparer);
+        }
+        return hash.ToHashCode();
+    }
+
+    private static IEnumerable<TElement> Snapshot(IEnumerable<TElement> source, ValueComparer<TElement> elementComparer)
+    {
+        // Deal with common cases first for performance
+        if (source.GetType().IsArray)
+        {
+            var sourceArray = (TElement[])source;
+            var snapshot = new TElement[sourceArray.Length];
+            for (int i = 0; i < sourceArray.Length; i++)
+                snapshot[i] = elementComparer.Snapshot(sourceArray[i]);
+            return snapshot;
+        }
+
+        if (source.GetType() == typeof(List<TElement>))
+        {
+            return ((List<TElement>)source).ConvertAll(elementComparer.Snapshot);
+        }
+
+        var constructors = source.GetType().GetConstructors().ToArray();
+
+        // Handle anything that has a constructor that accepts an IList<TElement>
+        if (HasConstructorWithSingleParameterOf<IList<TElement>>(constructors))
+        {
+            return CreateInstance(source.GetType(), source.Select(elementComparer.Snapshot).ToList());
+        }
+
+        // Handle anything that has a constructor that accepts an IEnumerable<TElement>
+        if (HasConstructorWithSingleParameterOf<IEnumerable>(constructors))
+        {
+            return CreateInstance(source.GetType(), source.Select(elementComparer.Snapshot));
+        }
+
+        // Handle anything with a default constructor and ICollection support to Add the items
+        if (source is ICollection<TElement> sourceCollection && constructors.Any(c => c.GetParameters().Length == 0))
+        {
+            var newCollection = (IList<TElement>)Activator.CreateInstance(source.GetType())!;
```

Comment:
Is this case tested?
Also do we need this if we are supporting only limited subset of collections?

Reply:
Was from prior more accepting code that failed in non-obvious ways, now removed.